CRM/Contribute/BAO/Contribution.php

```diff
@@ -4732,6 +4728,54 @@ public static function sendMail(&$input, &$ids, $contributionID, &$values, $recu
   }
 
   /**
+   * Generate From email and from name in an array values
+   */
+  public static function generateFromEmailAndName($input, $contribution) {
+    $domain = civicrm_api3('domain', 'getsingle', array('id' => CRM_Core_Config::domainID()));
+    $userID = CRM_Core_Session::singleton()->getLoggedInContactID();
+    $userName = '';
+    $userEmail = '';
+    if (!empty($userID)) {
+      list($userName, $userEmail) = CRM_Contact_BAO_Contact_Location::getEmailDetails($userID);
+    }
+    $fromName = '';
+    $fromEmail = '';
+    $pageValues = array();
+    if (!empty($contribution->contribution_page_id)) {
+      $pageValues = civicrm_api3('ContributionPage', 'getsingle', array('id' => $contribution->contribution_page_id));
+    }
+    // Use the values supplied to the function from the API.
+    if (!empty($input['receipt_from_name'])) {
```

Comment:
There is a logic issue - name & email are a set. If email is set then use the associated name, even if empty